python/mxnet/module/base_module.py

```diff
@@ -333,9 +334,13 @@ def fit(self, train_data, eval_data=None, eval_metric='acc',
 
         if hasattr(train_data, 'layout_mapper'):
             self.layout_mapper = train_data.layout_mapper
-        data_types = dict(train_data.provide_data_type + train_data.provide_label_type)
-        self.bind(data_shapes=train_data.provide_data, label_shapes=train_data.provide_label,
-                  for_training=True, force_rebind=force_rebind, input_types=data_types)
+
+        provide_data = DataDesc.get_list(train_data.provide_data,
+                                         train_data.provide_data_type)
+        provide_label = DataDesc.get_list(train_data.provide_label,
```

Comment:
since you can do `for name, shape in DataDesc`, you can make dataiter return DataDesc for provide_data, then you don't need provide_data_type anymore